### PR TITLE
Enable PgUp/PgDown and Home/End in the command palette

### LINKED ISSUE #7729: Command Palette should support page up/down to move faster thru the options
# Description of the new feature/enhancement
After pressing Ctrl+Shift+P, the only way to navigate thru the options is to use the key up/down. If I press Page Up or Page Down, it should move faster thru the options. Right now, I have to press the down key many times, or leave it pressed, which is far from an optimal experience.

# Proposed technical implementation details (optional)
I don't believe we need this to be customizable, so a fixed "page down/up" should just go to the end of the visible list, showing more options available, similar to how XAML's ListView already handles this scenario.

### PR DESCRIPTION
**Pull request by Hegunumo**
Closes #7729

### CODE REVIEW

**Comment by DHowett** [comment]
Hey there! You can keep updating the same pull request, you don’t have to make a new one every time! :smile: it might be a little easier for you to manage that way!

**Comment by zadjii-msft** [comment]
My only real qualm with this PR is hardcoding the height to 42. I'd really rather _not_ do that, and use the `ActualHeight` of the individual `ListViewItem` at runtime. However, I'm not positive that there's actually a way to do that. I don't know if it's possible to get a singular `ListViewItem` from a list view, which seems bizarre to me

**Comment by Hegunumo** [comment]
Yes I don't like it is being hardcoded either. I will try a little bit more.

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 1]
```diff
@@ -101,18 +101,21 @@ namespace winrt::TerminalApp::implementation
...
+        const auto listHeight = ::base::saturated_cast<int>(_filteredActionsView().ActualHeight());
+        const int numVisibleItems = listHeight / 42;
+
         const auto selected = _filteredActionsView().SelectedIndex();
         const int numItems = ::base::saturated_cast<int>(_filteredActionsView().Items().Size());
         // Wraparound math. By adding numItems and then calculating modulo numItems,
         // we clamp the values to the range [0, numItems) while still supporting moving
         // upward from 0 to numItems - 1.
-        const auto newIndex = ((numItems + selected + (moveDown ? 1 : -1)) % numItems);
+        const auto newIndex = ((numItems + selected + (moveDown ? (pageButtonPressed ? numVisibleItems : 1) : (pageButtonPressed ? -numVisibleItems : -1))) % numItems);
```
Comment (Don-Vito) [inline src/cascadia/TerminalApp/CommandPalette.cpp:118]: The logic duplication can be avoided by extracting a "delta" variable: moveDown? delta : -delta

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 2]
```diff
@@ -101,18 +101,21 @@ namespace winrt::TerminalApp::implementation
...
-    //   list. Otherwise, we're attempting to move to the previous.
+    // - moveDown: if true, we're attempting to move to the next or a few next item in the
+    //   list. Otherwise, we're attempting to move to the previous or not very many previous.
     // Return Value:
     // - <none>
-    void CommandPalette::SelectNextItem(const bool moveDown)
+    void CommandPalette::SelectNextItem(const bool moveDown, const bool pageButtonPressed)
     {
+        const auto listHeight = ::base::saturated_cast<int>(_filteredActionsView().ActualHeight());
+        const int numVisibleItems = listHeight / 42;
```
Comment (Don-Vito) [inline src/cascadia/TerminalApp/CommandPalette.cpp:111]: If 42 cannot be computed, consider extracting it to a const and documenting why this value was chosen

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 3]
```diff
@@ -101,18 +101,21 @@ namespace winrt::TerminalApp::implementation
     // - Moves the focus up or down the list of commands. If we're at the top,
     //   we'll loop around to the bottom, and vice-versa.
     // Arguments:
-    // - moveDown: if true, we're attempting to move to the next item in the
-    //   list. Otherwise, we're attempting to move to the previous.
+    // - moveDown: if true, we're attempting to move to the next or a few next item in the
+    //   list. Otherwise, we're attempting to move to the previous or not very many previous.
```
Comment (Don-Vito) [inline src/cascadia/TerminalApp/CommandPalette.cpp:105]: Document the new parameter

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 4]
```diff
@@ -161,13 +164,25 @@ namespace winrt::TerminalApp::implementation
...
+        else if (key == VirtualKey::PageUp)
+        {
+            // Action Mode: Move focus to the previous item in the list.
+            SelectNextItem(false, true);
+            e.Handled(true);
+        }
+        else if (key == VirtualKey::PageDown)
+        {
+            // Action Mode: Move focus to the previous item in the list.
+            SelectNextItem(true, true);
```
Comment (Don-Vito) [inline src/cascadia/TerminalApp/CommandPalette.cpp:185]: The original ticket also suggested support for Home and End. Consider implementing them as well :) In this case probably it makes sense to change the prototype of the SelectNextItem to receive enum.

Comment (Hegunumo) [reply src/cascadia/TerminalApp/CommandPalette.cpp:185]: I was forgetting about a support for Home and End. Yeah enum would be do it  for this.

**Comment by Hegunumo** [comment]
I've been trying to generate `ListViewItem` in CommandPalette.xaml.g.h like _filteredActionView() is so that I can access height property of `ListViewItem`, but I keep failing it. Though it's possible right ?

**Comment by DHowett** [comment]
I wonder if we can get rid of the 42 by using... `ListView::ContainerFromIndex()` and measuring the object it gives us to get its `ActualHeight()`?

**Comment by Hegunumo** [comment]
I have no idea how to measure the object that I got from `ListView::ContainerFromIndex()`. I have been trying to get top y coordinate and bottom y coordinate of the object and subtract one from another, but I don't know how to do it. Or is there specific API to measure the height of it ?

**Comment by Hegunumo** [comment]
Is this commit better than before? At least hardcoded 42 is removed.

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 5]
```diff
@@ -101,22 +101,43 @@ namespace winrt::TerminalApp::implementation
...
-    // - moveDown: if true, we're attempting to move to the next item in the
-    //   list. Otherwise, we're attempting to move to the previous.
+    // - moveDown: if true, we're attempting to move to the next or a few next item in the
+    //   list. Otherwise, we're attempting to move to the previous or not very many previous.
     // Return Value:
     // - <none>
-    void CommandPalette::SelectNextItem(const bool moveDown)
+    void CommandPalette::SelectNextItem(const bool moveDown, const bool pageButtonPressed)
     {
+        const auto itemHeight = ::base::saturated_cast<int>(_parentCommandText().Height());
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:110]: I wouldn't take a dependency on the height of the `parentCommandText` element, there's not necessarily a guarantee that element is the same height as the items of the list view

File: src/cascadia/TerminalApp/CommandPalette.xaml [hunk 6]
```diff
@@ -173,6 +173,7 @@ the MIT License. See LICENSE in the project root for license information. -->
 
             <TextBlock
                 Padding="16, 0, 16, 4"
+                Height="42"
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.xaml:176]: We _should_ be able to do this without manually setting a height

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
What happened to trying `ContainerFromIndex`? Does the following work to get the item's height?
```c++

        auto f = _filteredActionsView().ContainerFromIndex(0);
        auto g = f.try_as<winrt::Windows::UI::Xaml::Controls::ListViewItem>();
        auto h = g.ActualHeight();
```

I _think_ that should work, but `g` might end up being null. If it is, I'd maybe try casting it to a `FrameworkElement` instead

**Review by zadjii-msft — APPROVED** [review]
As much as I hate this ternary, I'd rather get this in for now and file a follow-up PR to clean it up. Someone else on the team can override that opinion though. Thanks for your patience!

File: src/cascadia/TerminalApp/App.xaml [hunk 7]
```diff
@@ -1,51 +1,59 @@
...
                     </Style>
 
-                    <!-- We need to manually create the error text brush as a
-                    theme-dependent brush. SystemControlErrorTextForegroundBrush
-                    is unfortunately static. -->
+                    <!--
+                        We need to manually create the error text brush as a
+                        theme-dependent brush. SystemControlErrorTextForegroundBrush
+                        is unfortunately static.
+                    -->
```
Comment (DHowett) [inline src/cascadia/TerminalApp/App.xaml:43]: I don't totally understand why this file changed

**Review by DHowett — APPROVED** [review]
Thanks for the contribution! This is great. I'm going to address my own comment about the App.xaml file and merge it.

**Comment by DHowett** [comment]
This is excellent work. :smile:

File: src/cascadia/TerminalApp/App.xaml (hunk 7 above)
Comment (Hegunumo) [reply src/cascadia/TerminalApp/App.xaml]: Yes. I don't either.

**Pull request merged**